external/fv3fit/fv3fit/pytorch/recurrent/generator.py

```diff
@@ -0,0 +1,401 @@
+from vcm.grid import get_grid_xyz
+
+
+@dataclasses.dataclass
+class RecurrentGeneratorConfig:
+    """
+    Configuration for a recurrent generator network.
+
+    Follows the architecture of Zhu et al. 2017, https://arxiv.org/abs/1703.10593.
+    This network contains an initial convolutional layer with kernel size 7,
+    strided convolutions with stride of 2, multiple residual blocks,
+    fractionally strided convolutions with stride 1/2, followed by an output
+    convolutional layer with kernel size 7 to map to the output channels.
+
+    Attributes:
+        n_convolutions: number of strided convolutional layers after the initial
+            convolutional layer and before the residual blocks
+        n_resnet: number of residual blocks
+        kernel_size: size of convolutional kernels in the strided convolutions
+            and resnet blocks
+        max_filters: maximum number of filters in any convolutional layer,
+            equal to the number of filters in the final strided convolutional layer
+            and in the resnet blocks
+        use_geographic_bias: if True, include a layer that adds a trainable bias
+            vector that is a function of x and y to the input and output of the network
+        samples_per_day: number of samples per model day, if given will provide the
+            model with two internal features tracking the position of the hour hand on
+            a 24 hour clock assuming the starting time of each window is midnight
+    """
+
+    n_convolutions: int = 3
+    n_resnet: int = 3
+    kernel_size: int = 3
+    strided_kernel_size: int = 4
+    max_filters: int = 256
+    use_geographic_bias: bool = True
+    use_geographic_features: bool = True
+    step_type: str = "resnet"
+    samples_per_day: Optional[int] = None
+
+    def build(
+        self,
+        channels: int,
+        nx: int,
+        ny: int,
+        n_time: int,
+        convolution: ConvolutionFactory = single_tile_convolution,
+    ):
+        """
+        Args:
+            channels: number of input channels
+            nx: number of x grid points
+            ny: number of y grid points
+            n_time: number of timesteps in output timeseries, including input timestep
+            convolution: factory for creating all convolutional layers
+                used by the network
+        """
+        return RecurrentGenerator(
+            config=self,
+            channels=channels,
+            nx=nx,
+            ny=ny,
+            n_time=n_time,
+            convolution=convolution,
+            step_type=self.step_type,
+        ).to(DEVICE)
+
+
+class PersistContext(nn.Module):
+    def __init__(self, op: nn.Module, context_channels: int):
+        super(PersistContext, self).__init__()
+        self.op = op
+        self.context_channels = context_channels
+
+    def forward(self, inputs: torch.Tensor) -> torch.Tensor:
+        """
+        Args:
+            inputs: tensor of shape [batch, tile, channels, x, y]
```

Comment:
I think this would be easy to rewrite if it ever changed. I don't expect this shape to ever change for this part of the code.

I've written very shape-independent code before which ended up very complex. What I think I've learned is that assuming shapes can make the code way simpler, we often don't need to be able to change the shape, and if we do it's easy enough to edit the code to change it.

I do at least not assume a time dimension exists for objects that don't need one, so you could create FoldFirstDimension(obj) if you want to apply it on a timeseries tensor. This could be used here but I don't see a need, as I don't think we'll ever use plain [channels, x, y] data.